### PR TITLE
FIX #30874: Feature request: Unable to add multiple emails at once as members to a room

### LINKED ISSUE #30874: [HOLD][$500] Feature request: Unable to add multiple emails at once as members to a room #26242
Coming from the below, which should be very similar
- https://github.com/Expensify/App/issues/26242 

If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to staging.new.expensify.com
2. Log in with any account
3. Add multiple emails at once as members to a room

## Expected Result:
Allow for multiple emails addresses to be added at once via a comma-separated list when inviting new members to a room.

## Actual Result:
Unable to add multiple emails at once as members to a workspace

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:**  1.3.95-5
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

image (20)

**Expensify/Expensify Issue URL:**
**Issue reported by:** @mallenexpensify 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1692837155454539

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~017b2a4652d665e512</li>
        <li>Upwork Job ID: 1699456983730249728</li>
        <li>Last Price Increase: 2023-11-06</li>
    </ul>
</details>

### PR DESCRIPTION
**Pull request by Samueljh1**
@mallenexpensify @mananjadhav 

### Details
Support was added to invite multiple users to a room simultaneously by supplying a comma-separated list of invitees.

### Fixed Issues
$ https://github.com/Expensify/App/issues/30874
PROPOSAL: https://github.com/Expensify/App/issues/30874#issuecomment-1799941663

### Tests
1. Open the room details for a room you own.
2. Click 'Members'.
3. Click 'Invite'.
4. Supply a comma-separated list of invite details, and ensure all valid entries are present in the search results.
5. Experiment with various valid/invalid emails in the list. Check error messages where (1. input is a single email with an invalid email, 2. input is a single email which is excluded (e.g., admin@expensify.com), 3. there are multiple emails supplied, none of which are valid).

- [X] Verify that no errors appear in the JS console

### Offline tests
Same steps as above.

### QA Steps
No platform-specific tests. Tested on all platforms.

- [X] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>
Android-Native

</details>

<details>
<summary>Android: mWeb Chrome</summary>
Android-Web

</details>

<details>
<summary>iOS: Native</summary>

iOS

</details>

<details>
<summary>iOS: mWeb Safari</summary>
iOS-Web

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/10816880/2b4337af-3dc6-4311-88ca-e61d76e72688

</details>

<details>
<summary>MacOS: Desktop</summary>
MacOS-Desktop

</details>

### CODE REVIEW

File: src/pages/RoomInvitePage.js [hunk 1]
```diff
@@ -82,22 +82,57 @@ function RoomInvitePage(props) {
...
+        // This ensures suggestions are displayed when there is no search term
+        // or the final list entry is empty
+        if (emails.length === 0 || cleanedSearchTerm.endsWith(',')) {
+            emails.push('');
+        }
+
+        const newUsersToInviteDict = {};
+        const newSelectedOptionsDict = {};
+
+        _.each(emails, (email, i) => {
```
Comment (mananjadhav) [inline src/pages/RoomInvitePage.js:100]: do we need the variable `i`? and if yes then can we rename it to `index`?

Comment (Samueljh1) [reply src/pages/RoomInvitePage.js:100]: Yes, it is used in the function body. I will rename

File: src/pages/RoomInvitePage.js [hunk 2]
```diff
@@ -82,22 +82,57 @@ function RoomInvitePage(props) {
         // eslint-disable-next-line react-hooks/exhaustive-deps
     }, []);
 
+    const cleanedSearchTerm = useMemo(() => searchTerm.trim().replace(/\s*,\s*/g, ','), [searchTerm]);
```
[the remark this answers is not in the document]
Comment (Samueljh1) [reply src/pages/RoomInvitePage.js:85]: This code is not present anywhere else. Will leave this for now.

**Comment by Samueljh1** [comment]
Any update @mananjadhav ?

**Comment by mananjadhav** [comment]
I will be reviewing this today/tomorrow.

**Comment by mananjadhav** [comment]
@Samueljh1 I tried testing this on Web, I can see the mixed emails still ignores the existing users in the room. It should throw an error like before.

https://github.com/Expensify/App/assets/3069065/4d9e9b9d-eae4-4c0e-a537-4a0c57fc46d6

and one may be unrelated to the PR (or even a valid bug). I notice some lag when I delete the characters from the input.

https://github.com/Expensify/App/assets/3069065/5899b8a9-019e-43a9-9439-36de8fb5f0db

**Comment by Samueljh1** [comment]
> @Samueljh1 I tried testing this on Web, I can see the mixed emails still ignores the existing users in the room. It should throw an error like before.
> 
>  web-room-invite-page.mov 
> and one may be unrelated to the PR (or even a valid bug). I notice some lag when I delete the characters from the input.
> 
>  web-room-invite-backspace.mov

Not sure the best approach to handle this, which is why I made it so there is only an error message when *no* users in the list are valid. Suppose we put ten emails in that are already members, do we clog up the screen with a huge error message listing all the invalid emails? This was intentional, but open to discussion..

**Comment by mananjadhav** [comment]
@Samueljh1 can you please close this PR as the issue is currently on hold?

**Pull request closed without merging**